# Innomatics Research Labs
### Internship September 2024
## Task 3 - Uber Fare Analysis

As a Data Analyst Intern with Uber.
###### TASK STATEMENT
Performm analysis and ML modelling on rides data recorded between 2009-01-01 and
2015-06-30.

### Import Libraries

In [542]:
import pandas as pd
import numpy as np

In [544]:
#Load Dataset
df=pd.read_excel("uber_rides_data.xlsx")

In [546]:
df.head()

ride_id  fare_amount          pickup_datetime  pickup_longitude  \
0  24238194          7.5  2015-05-07 19:52:06 UTC        -73.999817   
1  27835199          7.7  2009-07-17 20:04:56 UTC        -73.994355   
2  44984355         12.9  2009-08-24 21:45:00 UTC        -74.005043   
3  25894730          5.3  2009-06-26 08:22:21 UTC        -73.976124   
4  17610152         16.0  2014-08-28 17:47:00 UTC        -73.925023   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  
0        40.738354         -73.999512         40.723217                1  
1        40.728225         -73.994710         40.750325                1  
2        40.740770         -73.962565         40.772647                1  
3        40.790844         -73.965316         40.803349                3  
4        40.744085         -73.973082         40.761247                5

In [548]:
#What is the shape of given dataset?
df.shape

(200000, 8)

In [550]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   ride_id            200000 non-null  int64  
 1   fare_amount        200000 non-null  float64
 2   pickup_datetime    200000 non-null  object 
 3   pickup_longitude   200000 non-null  float64
 4   pickup_latitude    200000 non-null  float64
 5   dropoff_longitude  199999 non-null  float64
 6   dropoff_latitude   199999 non-null  float64
 7   passenger_count    200000 non-null  int64  
dtypes: float64(5), int64(2), object(1)
memory usage: 12.2+ MB


In [552]:
#How many integer columns(by default) are given in the dataset?
#Answer: 2 columns are integer by default i.e. ride_id and passenger_count

In [554]:
#Check how many missing values are in the 'dropoff_longitude' column
missing_values_count = df['dropoff_longitude'].isnull().sum()
print(f"Number of missing values in 'dropoff_longitude': {missing_values_count}")

Number of missing values in 'dropoff_longitude': 1


In [556]:
#How many missing values exists in 'dropoff_longitude' column?
#Answer: There is only 1 missing value in "dropoff_longitute" column.

In [558]:
df = df.dropna(subset=['dropoff_longitude'])

In [560]:
missing_values_count = df['dropoff_longitude'].isnull().sum()
print(f"Number of missing values in 'dropoff_longitude': {missing_values_count}")

Number of missing values in 'dropoff_longitude': 0


In [562]:
df = df.dropna(subset=['dropoff_latitude'])
missing_values_count = df['dropoff_latitude'].isnull().sum()
print(f"Number of missing values in 'dropoff_latitude': {missing_values_count}")

Number of missing values in 'dropoff_latitude': 0


In [564]:
#What is the datatpe of pickup_datetime column in the given dataset?
df['pickup_datetime'].dtype

dtype('O')

In [566]:
#what is correct syntax to convert pickup_datetime datatype from 'object' to 'datetime'?
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])
print(df.dtypes['pickup_datetime'])
# We can also use syntax-"df['pickup_datetime'].astype('datetime64[ns]')" 

datetime64[ns, UTC]


In [567]:
#Remove rows with null values in the 'fare_amount' column (and other relevant columns)
df_cleaned = df.dropna(subset=['fare_amount'])


In [571]:
#Calculate the average fare amount
average_fare = df_cleaned['fare_amount'].mean()

#Print the result
print(f"The average fare amount is: ${average_fare:.2f}")

The average fare amount is: $11.36


In [573]:
#What is the average fare amount?
#Answer: The average fare amount is: $11.36

In [575]:
null_values_count_fare_amount = df['fare_amount'].isnull().sum()
print(f"Number of null values count fare_amount: {null_values_count_fare_amount}")

Number of null values count fare_amount: 0


In [577]:
#Calculate distance between each pickup and dropoff points using Haversine formula. 
#What is the median haversine distance between pickup and dropoff location according to the given dataset?

# Haversine function to calculate the distance
def haversine(long1, lat1, long2, lat2):
    # Convert degrees to radians
    long1, lat1, long2, lat2 = map(np.radians, [long1, lat1, long2, lat2])
    
    # Haversine formula
    dlong = long2 - long1
    dlat = lat2 - lat1
    a = np.sin(dlat / 2.0)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlong / 2.0)**2
    c = 2 * np.arcsin(np.sqrt(a))
    
    # Earth's radius in kilometers (use 6371 for kilometers, 3959 for miles)
    r = 6371
    return c * r

# Apply the haversine formula to calculate the distance for each row
df_cleaned['haversine_distance'] = df_cleaned.apply(lambda row: haversine(row['pickup_longitude'], row['pickup_latitude'], row['dropoff_longitude'], row['dropoff_latitude']), axis=1)


In [579]:
# Calculate the median distance
median_distance = df_cleaned['haversine_distance'].median()

# Print the median Haversine distance
print(f"The median Haversine distance is: {median_distance:.2f} km")

The median Haversine distance is: 2.12 km


In [582]:
# Calculate the maximum distance
max_distance = df_cleaned['haversine_distance'].max()

# Print the maximum Haversine distance
print(f"The maximum Haversine distance is: {max_distance:.2f} km")

The maximum Haversine distance is: 16409.24 km


In [584]:
#How many rides have 0.0 haversine distance between pickup and dropoff location according to the given dataset?

In [586]:
# Count the number of rides where the Haversine distance is 0.0
zero_distance_rides = df_cleaned[df_cleaned['haversine_distance'] == 0.0].shape[0]

# Print the number of rides with 0.0 Haversine distance
print(f"The number of rides with 0.0 Haversine distance is: {zero_distance_rides}")

The number of rides with 0.0 Haversine distance is: 5632


In [588]:
#What is the mean 'fare_amount' for rides with 0 haversine distance?
#Do you sense something fishy? Try to analyze, and give your expert opinion in Jupyter Notebook.

In [590]:
df_cleaned.head()

ride_id  fare_amount           pickup_datetime  pickup_longitude  \
0  24238194          7.5 2015-05-07 19:52:06+00:00        -73.999817   
1  27835199          7.7 2009-07-17 20:04:56+00:00        -73.994355   
2  44984355         12.9 2009-08-24 21:45:00+00:00        -74.005043   
3  25894730          5.3 2009-06-26 08:22:21+00:00        -73.976124   
4  17610152         16.0 2014-08-28 17:47:00+00:00        -73.925023   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.738354         -73.999512         40.723217                1   
1        40.728225         -73.994710         40.750325                1   
2        40.740770         -73.962565         40.772647                1   
3        40.790844         -73.965316         40.803349                3   
4        40.744085         -73.973082         40.761247                5   

   haversine_distance  
0            1.683323  
1            2.457590  
2            5.036377  
3            1.661683  
4            4.475450

In [592]:
# Create a DataFrame for rides with 0.0 Haversine distance
zero_distance_rides = df_cleaned[df_cleaned['haversine_distance'] == 0.0]

# Display the first few rows of the new DataFrame (optional)
print(zero_distance_rides.head())

     ride_id  fare_amount           pickup_datetime  pickup_longitude  \
5   44470845          4.9 2011-02-12 02:27:09+00:00        -73.969019   
7   44195482          2.5 2012-12-11 13:52:00+00:00          0.000000   
11   6379048          8.5 2011-05-23 22:15:00+00:00          0.000000   
48  22405517         56.8 2013-01-03 22:24:41+00:00        -73.993498   
65  21993993          6.0 2014-05-05 19:27:00+00:00          0.000000   

    pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
5         40.755910         -73.969019         40.755910                1   
7          0.000000           0.000000          0.000000                1   
11         0.000000           0.000000          0.000000                1   
48        40.764686         -73.993498         40.764686                1   
65         0.000000           0.000000          0.000000                1   

    haversine_distance  
5                  0.0  
7                  0.0  
11                 0.0 

In [594]:
# Calculate the mean fare amount for rides with 0.0 Haversine distance
mean_fare_zero_distance = zero_distance_rides['fare_amount'].mean()

# Print the mean fare amount
print(f"The mean fare amount for rides with 0.0 Haversine distance is: ${mean_fare_zero_distance:.2f}")

The mean fare amount for rides with 0.0 Haversine distance is: $11.59


In [597]:
#What is the maximum 'fare_amount' for a ride?
df['fare_amount'].max()

499.0

In [599]:
#What is the haversine distance between pickup and dropoff location for the costliest ride?
#Do you sense something fishy? Try to analyze, and give your expert opinion in Jupyter Notebook.

In [601]:
#Find the row with the maximum fare amount
costliest_ride = df_cleaned.loc[df_cleaned['fare_amount'].idxmax()]

#Extract the pickup and dropoff coordinates for the costliest ride
pickup_long = costliest_ride['pickup_longitude']
pickup_lat = costliest_ride['pickup_latitude']
dropoff_long = costliest_ride['dropoff_longitude']
dropoff_lat = costliest_ride['dropoff_latitude']

#Calculate the Haversine distance for the costliest ride
costliest_ride_distance = haversine(pickup_long, pickup_lat, dropoff_long, dropoff_lat)

#Print the result
print(f"The Haversine distance for the costliest ride is: {costliest_ride_distance:.2f} km")

The Haversine distance for the costliest ride is: 0.00 km


In [603]:
#Answer: The result of 0.00 km for the Haversine distance of the costliest ride might indicate that the pickup and dropoff locations are the same.
#which is not acceptable for the costlier Uber ride.This is something suspicious!!
#This may b e due to data entry errors!

In [605]:
print(costliest_ride)

ride_id                                51151143
fare_amount                               499.0
pickup_datetime       2011-04-10 04:10:00+00:00
pickup_longitude                     -73.968377
pickup_latitude                       40.764602
dropoff_longitude                    -73.968368
dropoff_latitude                        40.7646
passenger_count                               1
haversine_distance                      0.00079
Name: 170081, dtype: object


In [607]:
#Print the result
print(f"The Haversine distance for the costliest ride is: {costliest_ride_distance:.5f} km")

The Haversine distance for the costliest ride is: 0.00079 km


In [609]:
df.head()

ride_id  fare_amount           pickup_datetime  pickup_longitude  \
0  24238194          7.5 2015-05-07 19:52:06+00:00        -73.999817   
1  27835199          7.7 2009-07-17 20:04:56+00:00        -73.994355   
2  44984355         12.9 2009-08-24 21:45:00+00:00        -74.005043   
3  25894730          5.3 2009-06-26 08:22:21+00:00        -73.976124   
4  17610152         16.0 2014-08-28 17:47:00+00:00        -73.925023   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  
0        40.738354         -73.999512         40.723217                1  
1        40.728225         -73.994710         40.750325                1  
2        40.740770         -73.962565         40.772647                1  
3        40.790844         -73.965316         40.803349                3  
4        40.744085         -73.973082         40.761247                5

In [611]:
#Extract the year from the 'pickup_datetime' column
df_cleaned['pickup_year'] = df_cleaned['pickup_datetime'].dt.year

#Filter the dataset for rides in the year 2014
rides_2014 = df_cleaned[df_cleaned['pickup_year'] == 2014]

#Count the number of rides in 2014
rides_2014_count = rides_2014.shape[0]

#Print the result
print(f"The number of rides recorded in 2014: {rides_2014_count}")

The number of rides recorded in 2014: 29968


In [613]:
#How many rides were recorded in the year 2014?
#Answer : 29968

In [615]:
# Extract year and month from 'pickup_datetime'
#df_cleaned['pickup_year'] = df_cleaned['pickup_datetime'].dt.year
df_cleaned['pickup_month'] = df_cleaned['pickup_datetime'].dt.month

# Filter the dataset for rides in the first quarter (January to March) of 2014
rides_first_quarter_2014 = df_cleaned[(df_cleaned['pickup_year'] == 2014) & (df_cleaned['pickup_month'].isin([1, 2, 3]))]

# Count the number of rides in the first quarter of 2014
rides_first_quarter_2014_count = rides_first_quarter_2014.shape[0]

# Print the result
print(f"The number of rides recorded in the first quarter of 2014: {rides_first_quarter_2014_count}")

The number of rides recorded in the first quarter of 2014: 7687


In [617]:
#How many rides were recorded in the first quarter of 2014?
#Answer: The number of rides recorded in the first quarter of 2014 is 7687.

In [619]:
#Filter for rides in September 2010
rides_september_2010 = df_cleaned[(df_cleaned['pickup_datetime'].dt.year == 2010) & (df_cleaned['pickup_datetime'].dt.month == 9)]

#Extract the day of the week (0 = Monday, 6 = Sunday)
rides_september_2010['day_of_week'] = rides_september_2010['pickup_datetime'].dt.dayofweek

#Count rides by day of the week
rides_by_day_of_week = rides_september_2010['day_of_week'].value_counts()

#Find the day with the maximum number of rides
max_rides_day = rides_by_day_of_week.idxmax()
max_rides_count = rides_by_day_of_week.max()

# Convert the numeric day of the week to the corresponding weekday name
days = {0: 'Monday', 1: 'Tuesday', 2: 'Wednesday', 3: 'Thursday', 4: 'Friday', 5: 'Saturday', 6: 'Sunday'}
max_rides_day_name = days[max_rides_day]

# Print the result
print(f"The day of the week with the maximum rides in September 2010 is {max_rides_day_name}, with {max_rides_count} rides.")

The day of the week with the maximum rides in September 2010 is Thursday, with 457 rides.


C:\Users\Dell\AppData\Local\Temp\ipykernel_14644\937782372.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rides_september_2010['day_of_week'] = rides_september_2010['pickup_datetime'].dt.dayofweek


In [621]:
#On which day of the week in September 2010, maximum rides were recorded ?
#Answer: The day of the week with the maximum rides in September 2010 is Thursday, with 457 rides

In [624]:
# Apply a Machine Learning Algorithm to predict the fare amount given following input features:
# passenger_count, distance and ride_week_day.

# Perform a 70-30 split of data.

# Which algorithm gives the least adjusted R square value?

##### Import Libraries

In [627]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import r2_score

In [629]:
df_cleaned['distance'] = df_cleaned.apply(lambda row: haversine(row['pickup_longitude'], row['pickup_latitude'], row['dropoff_longitude'], row['dropoff_latitude']), axis=1)

df_cleaned['ride_week_day'] = df_cleaned['pickup_datetime'].dt.dayofweek

In [630]:
df_cleaned.head()

ride_id  fare_amount           pickup_datetime  pickup_longitude  \
0  24238194          7.5 2015-05-07 19:52:06+00:00        -73.999817   
1  27835199          7.7 2009-07-17 20:04:56+00:00        -73.994355   
2  44984355         12.9 2009-08-24 21:45:00+00:00        -74.005043   
3  25894730          5.3 2009-06-26 08:22:21+00:00        -73.976124   
4  17610152         16.0 2014-08-28 17:47:00+00:00        -73.925023   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.738354         -73.999512         40.723217                1   
1        40.728225         -73.994710         40.750325                1   
2        40.740770         -73.962565         40.772647                1   
3        40.790844         -73.965316         40.803349                3   
4        40.744085         -73.973082         40.761247                5   

   haversine_distance  pickup_year  pickup_month  distance  ride_week_day  
0            1.683323         2015             5  1.683323              3  
1            2.457590         2009             7  2.457590              4  
2            5.036377         2009             8  5.036377              0  
3            1.661683         2009             6  1.661683              4  
4            4.475450         2014             8  4.475450              3

In [633]:
# Define the features and target
X = df_cleaned[['passenger_count', 'distance', 'ride_week_day']]
y = df_cleaned['fare_amount']

In [635]:
# Split data into training and testing (70-30 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [637]:

# Check for missing values in each column
missing_values_per_column = df_cleaned.isna().sum()

# Print the result
print(missing_values_per_column)

ride_id               0
fare_amount           0
pickup_datetime       0
pickup_longitude      0
pickup_latitude       0
dropoff_longitude     0
dropoff_latitude      0
passenger_count       0
haversine_distance    0
pickup_year           0
pickup_month          0
distance              0
ride_week_day         0
dtype: int64


In [639]:
# Train Linear Regression Model
linear_reg = LinearRegression()
linear_reg.fit(X_train, y_train)

LinearRegression()

In [641]:
# Predict using the model
y_pred_lr = linear_reg.predict(X_test)

In [643]:
# Calculate R-squared
r2_lr = r2_score(y_test, y_pred_lr)

In [645]:
# Manually calculate Adjusted R-squared
n = len(y_test)  # number of samples
p = X_test.shape[1]  # number of features
adj_r2_lr = 1 - (1 - r2_lr) * (n - 1) / (n - p - 1)

In [647]:
# Print R-squared and Adjusted R-squared
print(f"Linear Regression R-squared: {r2_lr:.4f}")
print(f"Linear Regression Adjusted R-squared: {adj_r2_lr:.4f}")

Linear Regression R-squared: 0.0004
Linear Regression Adjusted R-squared: 0.0004


In [649]:
# Decision Tree Classifier
#Decision Tree Classifier isn't suitable for predicting continuous values.(like fare amount)
#Similarly, KNN(K-Nearest Neighbour) isn't suitable for continuous values

In [653]:
# Train Random Forest Regression model
rf_reg = RandomForestRegressor()
rf_reg.fit(X_train, y_train)

# Predict using the model
y_pred_rf = rf_reg.predict(X_test)

# Calculate R-squared
r2_rf = r2_score(y_test, y_pred_rf)

# Manually calculate Adjusted R-squared
adj_r2_rf = 1 - (1 - r2_rf) * (n - 1) / (n - p - 1)

# Print R-squared and Adjusted R-squared
print(f"Random Forest Regression R-squared: {r2_rf:.4f}")
print(f"Random Forest Regression Adjusted R-squared: {adj_r2_rf:.4f}")

Random Forest Regression R-squared: 0.6311
Random Forest Regression Adjusted R-squared: 0.6311


###### The Random Forest Regression is better in this case based on the R-squared and Adjusted R-squared values.
###### Linear Regression model explains only about 0.04% of the variance in the fare amounts, which is extremely low.
###### The Random Forest Regression model explains about 63.11% of the variance in the fare amounts, which is a far better result.